### PR TITLE
Refactor some common utilities into simutils

### PR DESCRIPTION
**Pull request by MTCam**
* Adds parallel grid generator common to all examples
* Generalizes checkpoint to be used even for sims without exact soln (needed for Y0/Y1 )
* Adds a placeholder for profiling util (also used in Y0/Y1 example)

Note: Y0/Y1 Euler example will follow in a separate PR

### CODE REVIEW

File: examples/lump-mpi.py [hunk 1]
```diff
@@ -85,34 +86,10 @@ def main(ctx_factory=cl.create_some_context):
...
-
-        part_per_element = get_partition_by_pymetis(mesh, num_parts)
-
-        local_mesh = mesh_dist.send_mesh_parts(mesh, part_per_element, num_parts)
-        del mesh
-
-    else:
-        local_mesh = mesh_dist.receive_mesh_part()
+    local_mesh, global_nelements = create_parallel_box_grid(comm, dim, box_ll,
+                                                            box_ur, nel_1d)
```
Comment (inducer) [inline examples/lump-mpi.py:92]: I think I'd prefer something a bit more general that gets the mesh by a callback. (rather than being specific to boxes)

Comment (MTCam) [reply examples/lump-mpi.py]: Yeah, ok, try this on.   I *think* this is something like what you are getting at.

File: mirgecom/simutil.py [hunk 2]
```diff
@@ -104,46 +104,94 @@ def exact_sim_checkpoint(discr, exact_soln, visualizer, eos, q,
...
         return 0
 
-    actx = q[0].array_context
-    nodes = thaw(actx, discr.nodes())
-    rank = 0
-
-    if comm is not None:
-        rank = comm.Get_rank()
+    if profiler is not None:
+        profiler.StartTimer(f"checkpoint_{step}")
```
Comment (inducer) [inline mirgecom/simutil.py:108]: I'd be more comfortable leaving these out. For one, this seems to preempt https://github.com/illinois-ceesd/planning/issues/57, and for another, the non-PEP8 interface sticks out to me like a sore thumb.

File: mirgecom/simutil.py [hunk 3]
```diff
@@ -104,46 +104,94 @@ def exact_sim_checkpoint(discr, exact_soln, visualizer, eos, q,
...
-            ("dv", dependent_vars),
-            ("exact_soln", expected_state),
-            ("residual", q - expected_state)
+            ("cv", cv),
+            ("dv", dependent_vars)
         ]
+        if exact_soln is not None:
+            exact_list = [
+                ("exact_soln", expected_state),
+                ("residual", exp_resid)
```
Comment (inducer) [inline mirgecom/simutil.py:141]: I'm kind of skeptical about adding this to the dump. At least paraview can straightforwardly calculate this internally.

File: mirgecom/simutil.py [hunk 4]
```diff
@@ -104,46 +104,94 @@ def exact_sim_checkpoint(discr, exact_soln, visualizer, eos, q,
...
-        if maxerr > exittol:
-            raise ExactSolutionMismatch(step, t=t, state=q)
+    if profiler is not None:
+        profiler.EndTimer(f"checkpoint_{step}")
+
+    if maxerr > exittol:
+        raise ExactSolutionMismatch(step, t=t, state=q)
+
+
+def create_parallel_grid(comm, grid_generator):
```
Comment (inducer) [inline mirgecom/simutil.py:174]: Documentation?

Comment (inducer) [inline mirgecom/simutil.py:174]: Callables should be named with verbs.

File: mirgecom/simutil.py (hunk 2 above)
Comment (MTCam) [reply mirgecom/simutil.py:108]: This is removed now, however...

I don't think it duplicates https://github.com/illinois-ceesd/planning/issues/57.   That Issue does not mention timing user-defined constructs - which I need.   Timing kernels only does not give the information that I need to profile the simulation.   For example, how do I use anything from https://github.com/illinois-ceesd/planning/issues/57 to time the I/O?  What if I want to time how long I spend inside the checkpoint routine (which I do)?

File: mirgecom/simutil.py (hunk 3 above)
Comment (MTCam) [reply mirgecom/simutil.py:141]: It was only convenience.  It is gone.

File: mirgecom/simutil.py (hunk 4 above)
Comment (MTCam) [reply mirgecom/simutil.py:174]: Added - but possibly all messed up, heh.  We'll see.

Comment (MTCam) [reply mirgecom/simutil.py:174]: Fixed.

Comment (MTCam) [reply mirgecom/simutil.py]: I actually think it is OK, except I didn't figure out how to document the MPI communicator in a great way.

Comment (inducer) [reply mirgecom/simutil.py]: See 83bb205.

File: mirgecom/simutil.py (hunk 2 above)
Comment (inducer) [reply mirgecom/simutil.py]: (see the discussion there)

**Review by inducer — APPROVED** [review]

**Pull request merged**